src/module.c

```diff
@@ -5442,7 +5442,10 @@ int moduleTimerHandler(struct aeEventLoop *eventLoop, long long id, void *client
             raxRemove(Timers,(unsigned char*)ri.key,ri.key_len,NULL);
             zfree(timer);
         } else {
-            next_period = (expiretime-now)/1000; /* Scale to milliseconds. */
+            /* We call ustime() again instead of using the cached 'now' so that
+             * 'next_period' isn't affected by the time it took to execute
+             * previous calls to 'callback. */
+            next_period = (expiretime-ustime())/1000; /* Scale to milliseconds. */
```

Comment:
If it is to improve the accuracy of the timer, I think it is better to put the update of now directly into the while loop. For example, there are multiple timers at the same time. After the callback of the previous timer is executed (it may take a long time to execute), the expiretime of the next timer has arrived, but in the current implementation, you need to execute this timer in the next event callback.

Reply:
if we update `now` inside the loop and the timer callback takes less than the period we can get stuck in an endless loop... 
example: my timer registers itself t run every 1s and takes 2s to execute.
let's say that on the first iteration now=9000 and expiretime=8999 so we execute the callback which takes 2s and re-registers timer for 1s from now which is 9001
next loop now=9002 and expiretime is 9001 so the callback is executed again, endlessly

Reply:
i agree with the current code of the PR, the criteria for triggering all timers should be the time in which the loop started.
we can maybe add complicated code to better serve other timers and still avoid endless loop, but i'm not sure the effort is worth it.
this fix is still simple, and just solves the bug which would cause a timer's request to be correctly recorded in the rax, but the loop exiting and telling redis to sleep for too long.